File: src/main/java/com/wesker926/leetcode/algorithms/p0869/s2/Solution.java
package com.wesker926.leetcode.algorithms.p0869.s2;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * @author wesker.gh
 * @date 2021/10/28
 * @description sort and compare
 */
public class Solution {

    private final Set<String> set;

    {
        set = new HashSet<>();
        for (int i = 1; i <= 1000000000; i <<= 1) {
            set.add(numToStr(i));
        }
    }

    public boolean reorderedPowerOf2(int n) {
        return set.contains(numToStr(n));
    }

    private String numToStr(int n) {
        char[] cs = String.valueOf(n).toCharArray();
        Arrays.sort(cs);
        return new String(cs);
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0869/s2/solution.go
// Package s2
// @author: wesker.gh
// @date: 2021/10/28
// @description: sort and compare
package s2

import (
	"sort"
	"strconv"
)

var cache = initCache()

func initCache() map[string]interface{} {
	c, n := make(map[string]interface{}), 1
	for ; n <= 1000000000; n <<= 1 {
		c[numToStr(n)] = nil
	}
	return c
}

func reorderedPowerOf2(n int) bool {
	_, ok := cache[numToStr(n)]
	return ok
}

func numToStr(n int) string {
	r := []byte(strconv.Itoa(n))
	sort.Slice(r, func(i, j int) bool {
		if r[i] < r[j] {
			return true
		}
		return false
	})
	return string(r)
}
